core/arch/arm/kernel/ree_fs_ta.c

```diff
@@ -0,0 +1,335 @@
+
+	/* Request TA from tee-supplicant */
+	res = rpc_load(uuid, &ta, &cookie, &ta_size);
+	if (res != TEE_SUCCESS)
+		goto error;
+
+	/* Make secure copy of signed header */
+	res = alloc_and_copy_shdr(&shdr, ta, ta_size);
+	if (res != TEE_SUCCESS)
+		goto error_free_payload;
+
+	/* Validate header signature */
+	res = check_shdr(shdr);
+	if (res != TEE_SUCCESS)
+		goto error_free_payload;
+
+	/*
+	 * Initialize a hash context and run the algorithm over the signed
+	 * header (less the final file hash and its signature of course)
+	 */
+	hash_algo = TEE_DIGEST_HASH_TO_ALGO(shdr->algo);
+	res = crypto_ops.hash.get_ctx_size(hash_algo, &hash_ctx_size);
+	if (res != TEE_SUCCESS)
+		goto error_free_payload;
+	hash_ctx = malloc(hash_ctx_size);
+	if (!hash_ctx) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto error_free_payload;
+	}
+	res = crypto_ops.hash.init(hash_ctx, hash_algo);
+	if (res != TEE_SUCCESS)
+		goto error_free_payload;
+	res = crypto_ops.hash.update(hash_ctx, hash_algo, (uint8_t *)shdr,
+				     sizeof(*shdr));
+	if (res != TEE_SUCCESS)
+		goto error_free_payload;
+
+	if (ta_size != SHDR_GET_SIZE(shdr) + shdr->img_size) {
+		res = TEE_ERROR_SECURITY;
+		goto error_free_payload;
+	}
+
+	handle->nw_ta = ta;
+	handle->nw_ta_size = ta_size;
+	handle->cookie = cookie;
+	handle->offs = SHDR_GET_SIZE(shdr);
+	handle->hash_algo = hash_algo;
+	handle->hash_ctx = hash_ctx;
+	handle->shdr = shdr;
+	*h = handle;
+	return TEE_SUCCESS;
+
+error_free_payload:
+	thread_rpc_free_payload(cookie);
+error:
+	free(hash_ctx);
+	free(shdr);
+	free(handle);
+	return res;
+}
+
+static TEE_Result ta_get_size(const struct user_ta_store_handle *h,
+			      size_t *size)
+{
+	*size = h->shdr->img_size;
+	return TEE_SUCCESS;
+}
+
+static TEE_Result check_digest(struct user_ta_store_handle *h)
+{
+	void *digest = NULL;
+	TEE_Result res;
+
+	if (!crypto_ops.hash.final)
+		return TEE_ERROR_NOT_SUPPORTED;
+	digest = malloc(h->shdr->hash_size);
+	if (!digest)
+		return TEE_ERROR_OUT_OF_MEMORY;
+	res = crypto_ops.hash.final(h->hash_ctx, h->hash_algo, digest,
+				    h->shdr->hash_size);
+	if (res != TEE_SUCCESS) {
+		res = TEE_ERROR_SECURITY;
+		goto out;
+	}
+	if (memcmp(digest, SHDR_GET_HASH(h->shdr), h->shdr->hash_size))
+		res = TEE_ERROR_SECURITY;
+out:
+	free(digest);
+	return res;
+}
+
+static TEE_Result ta_read(struct user_ta_store_handle *h, void *data,
+			  size_t len)
+{
+	uint8_t *src = (uint8_t *)h->nw_ta + h->offs;
+	uint8_t *dst = src;
+	TEE_Result res;
+
+	if (h->offs + len > h->nw_ta_size)
+		return TEE_ERROR_BAD_PARAMETERS;
+	if (data) {
+		dst = data; /* Hash secure buffer (shm might be modified) */
+		memcpy(dst, src, len);
+	}
+	res = crypto_ops.hash.update(h->hash_ctx, h->hash_algo, dst, len);
+	if (res != TEE_SUCCESS)
+		return TEE_ERROR_SECURITY;
+	h->offs += len;
+	if (h->offs == h->nw_ta_size) {
+		/*
+		 * Last read: time to check if our digest matches the expected
+		 * one (from the signed header)
+		 */
+		res = check_digest(h);
```

Comment:
This seems a bit fragile. What if there's some mismatch that prevents `h->offs` from reaching `h->nw_ta_size`?
I don't have a good idea how to fix this right now.